src/backend/resources.rs

```diff
@@ -0,0 +1,228 @@
+        _file: &mut OutputFile,
+    ) -> CalyxResult<()> {
+        Ok(())
+    }
+
+    fn emit(ctx: &ir::Context, file: &mut OutputFile) -> CalyxResult<()> {
+        let main_comp = ctx
+            .components
+            .iter()
+            .find(|comp| comp.name == ctx.entrypoint)
+            .unwrap();
+
+        let count_map: &mut HashMap<(Id, Binding), u32> = &mut HashMap::new();
+
+        gen_count_map(ctx, main_comp, count_map);
+
+        write_csv(count_map.clone(), file);
+
+        estimated_size(count_map.clone());
+
+        Ok(())
+    }
+}
+
+/// Counts the number of each primitive with a given set of parameters
+/// in the program with entrypoint `main_comp`.
+/// TODO: handle multi-component designs
+fn gen_count_map(
+    ctx: &ir::Context,
+    main_comp: &Component,
+    count_map: &mut HashMap<(Id, Binding), u32>,
+) {
+    for cell in main_comp.cells.iter() {
+        let cell_ref = cell.borrow();
+        let primitives = vec![
+            "std_reg",
+            "std_mem_d1",
+            "std_mem_d2",
+            "std_mem_d3",
+            "std_mem_d4",
+            "seq_mem_d1",
+            "seq_mem_d2",
+            "seq_mem_d3",
+            "seq_mem_d4",
+        ];
+        match &cell_ref.prototype {
+            CellType::Primitive {
+                name,
+                param_binding,
+                ..
+            } => {
+                if primitives.contains(&name.id.as_str()) {
+                    *count_map
+                        .entry((*name, (**param_binding).clone()))
+                        .or_insert(0) += 1;
+                }
+            }
+            CellType::Component { name } => {
+                let component = ctx
+                    .components
+                    .iter()
+                    .find(|comp| comp.name == name)
+                    .unwrap();
+                gen_count_map(ctx, component, count_map);
+            }
+            _ => (),
+        }
+    }
+}
+
+/// Writes a CSV to stdout with primitive count information
+/// generated by `gen_count_map`.
+/// TODO (priya): fix the output format and have it output to an actual file!
+fn write_csv(count_map: HashMap<(Id, Binding), u32>, _file: &mut OutputFile) {
+    let mut wtr = csv::Writer::from_writer(io::stdout());
+    let header = vec!["Primitive", "Count", "Parameters"];
+    wtr.write_record(header).ok();
+    wtr.write_record(None::<&[u8]>).ok();
+    for ((name, params), count) in count_map {
+        let mut result = String::new();
+        result.push_str(name.id.as_str());
+        result.push(',');
+        result.push_str(count.to_string().as_str());
+        for (id, val) in params {
+            result.push(',');
+            result.push_str(id.id.as_str());
+            result.push(',');
+            result.push_str(val.to_string().as_str());
+        }
+        result.push('\n');
+        let ret: Vec<&str> = result.split(',').collect();
+        wtr.write_record(ret).ok();
```

Comment:
TBH I don't quite follow why this is building up a string and then `split`ing it again on the commas in the string. Would it work to do something more like this?

```rust
let params_str = /* do whatever needs to be done to stringify the binding, preferably not using the , character */;
wtr.write_record(vec![name, count.to_string(), params_str]);
```

Reply:
thank you! I ended up doing this!